cli/command/stack/kubernetes/services.go

```diff
@@ -79,60 +78,47 @@ func getResourcesForServiceList(dockerCli *KubeCli, filters filters.Args, labelS
 	return replicas, daemons, services, nil
 }
 
-// RunServices is the kubernetes implementation of docker stack services
-func RunServices(dockerCli *KubeCli, opts options.Services) error {
+// GetServices is the kubernetes implementation of listing stack services
+func GetServices(dockerCli *KubeCli, opts options.Services) ([]swarm.Service, map[string]service.ListInfo, error) {
 	filters := opts.Filter.Value()
 	if err := filters.Validate(supportedServicesFilters); err != nil {
-		return err
+		return nil, nil, err
 	}
 	client, err := dockerCli.composeClient()
 	if err != nil {
-		return nil
+		return nil, nil, err
 	}
 	stacks, err := client.Stacks(false)
 	if err != nil {
-		return nil
+		return nil, nil, err
 	}
 	stackName := opts.Namespace
 	_, err = stacks.Get(stackName)
 	if apierrs.IsNotFound(err) {
-		return fmt.Errorf("nothing found in stack: %s", stackName)
+		return []swarm.Service{}, nil, nil
```

Comment:
I was looking if we should return the error here, and have it handled by the caller 🤔. That way, this function (when used in different situations) would not hide this information.

I think that's ok for a follow-up, because we should have a function to convert Kubernetes API errors to `errdef` errors. In that case we would map (if possible) / standardise these errors, so that the non-k8s specific code understands them; https://github.com/docker/cli/blob/0b6685bca8383437ebaa635437db2891fc77711d/vendor/k8s.io/apimachinery/pkg/api/errors/errors.go#L456-L459

Similar to https://github.com/moby/moby/blob/81dbed4c8b87d99824eecea68c95f10056fd3d2a/errdefs/http_helpers.go#L172-L191

```go
// statusCodeFromContainerdError returns status code for containerd errors when
// consumed directly (not through gRPC)
func statusCodeFromContainerdError(err error) int {
	switch {
	case containerderrors.IsInvalidArgument(err):
		return http.StatusBadRequest
	case containerderrors.IsNotFound(err):
		return http.StatusNotFound
	case containerderrors.IsAlreadyExists(err):
		return http.StatusConflict
	case containerderrors.IsFailedPrecondition(err):
		return http.StatusPreconditionFailed
	case containerderrors.IsUnavailable(err):
		return http.StatusServiceUnavailable
	case containerderrors.IsNotImplemented(err):
		return http.StatusNotImplemented
	default:
		return http.StatusInternalServerError
	}
}
```

Reply:
The reason I didn't return an error here is because the same code for swarm returns an empty slice when no services are found. We can do it one way or the other, we should just make sure that errors are returned in the same places for kubernetes and swarm.